### PR TITLE
Add Base GEMM

### PR DESCRIPTION
**Pull request by xiaoling-yi**
This RP adds the Base GEMM:
* Add comments to the source and test file
* Add Unit test for the Base GEMM
* Add the golden model in Scala
* Add documentation for the Base GEMM
* Refactor the signal names
*  Add the accumulate input for accumulating partial output data

### CODE REVIEW

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
Hi @xiaoling-yi, this seems to remove all previously existing files and replace them with entirely new files, even though they are not completely new? Could you make the changes in the previously existing files please? Otherwise it kind of defeats the purpose of using Git

**Comment by xiaoling-yi** [comment]
Hi @JosseVanDelm , the file names have been changed to more format ones. The code is based on the previously existing files.

**Comment by JosseVanDelm** [comment]
I see, but now I can not properly review the changes. The diff just shows completely new files, and I can not see what you added/removed. Can we make a seperate PR for the refactoring of the names later please, that keeps a proper blame too?

**Comment by xiaoling-yi** [comment]
Hi @JosseVanDelm , Yes we can. How can we do this?

**Comment by JosseVanDelm** [comment]
Files are renamed now and ready for review!

File: src/main/scala/gemm/gemm.scala [hunk 1]
```diff
@@ -3,75 +3,93 @@ import chisel3._
...
+  io.data_valid_o := chek_data_in_valid_reg
 }
 
 // Mesh IO definition, an extended version of Tile IO
 class MeshIO extends Bundle {
-  val a_io_in = Input(
+  val a_i = Input(
     Vec(
       GEMMConstant.meshRow,
       Vec(GEMMConstant.tileSize, UInt(GEMMConstant.input.W))
```
Comment (JosseVanDelm) [inline src/main/scala/gemm/gemm.scala:76]: Why are we using `SInt` for the type of `c_o` but not for `a_i` or `b_i`?

File: src/test/scala/gemm/MatrixlibBase.scala [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      meshCol: Int,
+      A: Array[Int]
+  ) = {
+    for (i <- 0 until meshRow) {
+      for (j <- 0 until meshCol) {
+        println(i, j, A(i * meshCol + j))
+      }
+    }
+    println("print 1D matrix over!")
+  }
```
Comment (JosseVanDelm) [inline src/test/scala/gemm/MatrixlibBase.scala:61]: Why is this function here twice?

File: src/main/scala/gemm/Parameter.scala [hunk 3]
```diff
@@ -5,9 +5,9 @@ import chisel3.util._
 object GEMMConstant {
 
   def input = 8
```
Comment (JosseVanDelm) [inline src/main/scala/gemm/Parameter.scala:7]: Both your documentation and this parameter file mention the possibility to change the bitwidth of this number, but your test hardcodes these values inside the types (e.g. use `Byte` for inputs) for for the golden model.
I feel like we should either:
* add a disclaimer that only these datatypes are tested and supported (easy)
* add support for random datatypes in the golden model and tests (hard!)
I'd definitely stick with the first option for now (and in this PR please!)

File: src/test/scala/gemm/MatrixlibBase.scala [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+      B: Array[BigInt]
+  ) = {
+    for (i <- 0 until Len) {
+      // println(A(i), B(i))
+      assert(A(i) == B(i))
+    }
+  }
+
+  // Matrix multiplication with one dimension Array[Byte] data type
+  def MarixMul_1D(
```
Comment (JosseVanDelm) [inline src/test/scala/gemm/MatrixlibBase.scala:163]:
```suggestion
  def MatrixMul_1D(
```

File: src/test/scala/gemm/MatrixlibBase.scala [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+          ) + A(i * tileSize + k) * B(k + j * tileSize)
+        }
+        // println("golden",golden(i * meshCol + j))
+      }
+    }
+    golden
+  }
+
+  // Matrix multiplication with 2 dimension Array[Array[Byte]] data type
+  def MarixMul_2D(
```
Comment (JosseVanDelm) [inline src/test/scala/gemm/MatrixlibBase.scala:189]:
```suggestion
  def MatrixMul_2D(
```

File: src/test/scala/gemm/MatrixlibBase.scala [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            i * meshCol + j
+          ) + A(i * tileSize + k) * B(k + j * tileSize)
+        }
+        // println("golden",golden(i * meshCol + j))
+      }
+    }
+    golden
+  }
+
+  // Matrix multiplication with 2 dimension Array[Array[Byte]] data type
```
Comment (JosseVanDelm) [inline src/test/scala/gemm/MatrixlibBase.scala:188]: This function does not seem to be used?

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
Hi @xiaoling-yi this is a really good PR! It is going in the right direction!

Some major comments:
* You seem a bit inconsistent with your use of `SInt` and `UInt`? throughout your design.
* The description for your PR could be a bit more elaborate, because this one adds a lot of extra stuff: you add a golden model, unit tests, and an accumulate input! especially the golden model and accumulate input are mentioned nowhere in your PR description.
* You forgot to attach your schematics in this PR
* Your docs and parameters allow parameterizing your IP, but you only test for the default case, and the test itself is not parameterizable.
* You add a lot of commented out prints in your test. These are not available in the automated tester. I'm wondering whether we should remove these calls.

File: README.md [hunk 7]
```diff
@@ -1,19 +1,44 @@
...
-This directory contains the SystemVerilog controller for connecting the Gemm to the SNAX core.
-### tb
-This directory contains the testbench for the SystemVerilog version of Gemm and a simple testbench for Accelerator with the Gemm.
-### dv
-This directory contains the design verification utils for running the testbench for Gemm and the Accelerator.
-
-## Chisel environment
+## Microarchitecture
+The microarchitecture of the GEMM accelerator is shown below. The GEMM array has meshRow row and meshCol column tiles. Each tile implements a tileSize vector dot product.
+![](./docs/microarch.png)
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply README.md:8]: Added!

File: src/main/scala/gemm/gemm.scala [hunk 8]
```diff
@@ -3,75 +3,93 @@ import chisel3._
...
 
-  chek_data_in_valid := io.in_control.data_in_valid
+  chek_data_in_valid_reg := io.control_i.data_valid_i
 
+  // Element-wise multiply
   for (i <- 0 until GEMMConstant.tileSize) {
-    mul_add_result_vec_reg(i) := io.a_io_in(i) * io.b_io_in(i)
+    mul_add_result_vec(i) := io.a_i(i).asSInt * io
+      .b_i(i)
+      .asSInt
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply src/main/scala/gemm/gemm.scala:46]: It is strange but it is generated by scalamft.

File: src/test/scala/gemm/MatrixlibBase.scala (hunk 6 above)
Comment (xiaoling-yi) [reply src/test/scala/gemm/MatrixlibBase.scala]: removed!

File: src/main/scala/gemm/gemm.scala [hunk 9]
```diff
@@ -3,75 +3,93 @@ import chisel3._
...
+// When data_valid_i assert, do the computation, and give the result next cycle, with a data_valid_o assert
 class Tile extends Module {
   val io = IO(new TileIO())
 
-  val accumulation_reg = RegInit(0.U(GEMMConstant.acc.W))
-  val chek_data_in_valid = RegInit(0.B)
-  val mul_add_result = Wire(UInt(GEMMConstant.mul.W))
-  val mul_add_result_vec_reg = Wire(
-    Vec(GEMMConstant.tileSize, UInt(GEMMConstant.mul.W))
+  val accumulation_reg = RegInit(0.S(GEMMConstant.accumulate.W))
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply src/main/scala/gemm/gemm.scala:30]: Rename var to be more clear about the data type

File: src/main/scala/gemm/gemm.scala [hunk 10]
```diff
@@ -3,75 +3,93 @@ import chisel3._
...
+  val result_reg = RegInit(0.S(GEMMConstant.accumulate.W))
+  val chek_data_in_valid_reg = RegInit(0.B)
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply src/main/scala/gemm/gemm.scala]: Changed!

**Review by rgantonio — APPROVED** [review]
Nice work! Just an FYI, @xiaoling-yi  and @rgantonio discussed in person for speed up purposes 😄 

Issues were made for TODO tasks.

@xiaoling-yi please make the filename changes you mentioned earlier.

**Review by JosseVanDelm — APPROVED** [review]
Thanks @rgantonio !

**Pull request merged**